import_table.py

```diff
@@ -1,88 +1,166 @@
+    # Try to open the provided file name.
+    try:
+        reader = csv.DictReader(csv_file, delimiter=',')
+    except:
+        raise BaseException('Cannot read/import CSV file. Check format and resubmit.')
 
-with open(FILENAME, 'r') as csv_file:
-    #Attempt to Read file
+    # Verify that the file has all the attributes we need.
+    attributes = set(reader.fieldnames)
     try:
-        reader = csv.DictReader(csv_file, delimiter=",")
+        attributes.remove('Progress')
+        attributes.remove('Email')
     except:
-        raise BaseException("Cannot read/import csv file. Check format and resubmit")
-    
-    #Search through file for relevant user data
-    userNum = 0
-    for user in reader:
-        userData = {}
-        userData['role'] = []
-        
-        #Gather email and discord_username
-        for attr in user:
-
-            #Find Discord Username
-            if(attr == ATTR_DIS and user[attr] != ""):
-                userData['discord_username'] = user[attr].replace(" ","")
-
-            #Find Email Address
-            if(attr == ATTR_EMAIL and user[attr] != ""):
-                userData['email'] = user[attr].replace(" ", "")
-
-            #Find Role
-            if('participant' in FILENAME.lower()):
-                userData['role'].append('participant')
+        raise ValueError('CSV file missing required attributes. Check file contents and resubmit.')
+
+    # Check if we are importing the participant or volunteer form.
+    is_participant = False
+    data_attr = VOLUNTEER_DATA_ATTR
+    try:
+        attributes.remove('Roles')
+    except:
+        is_participant = True
+        data_attr = PARTICIPANT_DATA_ATTR
+
+    # For each entry in the CSV file...
+    for entry in reader:
+        num_entries = num_entries + 1
+
+        # Check that the entry is for a completed response.
+        if entry['Progress'] != '100':
+            num_unfinished = num_unfinished + 1
+            continue
+
+        # Check for and store the entry's email.
+        if entry['Email'] == '':
+            num_error = num_error + 1
+            continue
+        email = entry['Email'].replace(' ', '')
+
+        # Check for and store the entry's roles.
+        roles = []
+        if is_participant:
+            roles.append('participant')
+        else:
+            # If the roles attribute exists but is blank, skip this entry.
+            if entry['Roles'] == '':
+                num_error = num_error + 1
+                continue
+            # Add appropriate roles for the volunteer form.
+            if JUDGE_ROLE_NUM in entry['Roles']:
+                roles.append('judge')
+            if MENTOR_ROLE_NUM in entry['Roles']:
+                roles.append('mentor')
+
+        # Check for and store all non-essential data.
+        data = {}
+        for attribute in data_attr.keys():
+            try:
+                data[data_attr[attribute]] = entry[attribute]
+            except:
+                pass
+
```

Comment:
Lines 118 through 149 can basically be achieved with a single method call:
```
records.add_registered_user(email, roles, data)
```
This is because the method removes any entries that already exist in the database (DB). But simply calling `add_registered_user()` incurs a heavy performance penalty, since each duplicate entry would have a `DELETE` and `INSERT` query issued for it. Which would the majority of entries for our intended use case (importing missed entries after the bot goes offline).

To work around this, I first check if an entry exists in the DB. I then retrieve the existing version of the relevant entry (`old_entry`) and perform a series of checks to determine if the two entries are the same or not.